docs(core): add some more info about the "why" of the automated update (#17955)
nrwl/nx: docs/shared/core-features/automate-updating-dependencies.md

# Automate Updating Dependencies

Keeping a codebase updated with the latest changes in your framework of choice and tooling can be challenging. Not to mention that "tooling maintenance work" is usually hard to squeeze into your feature sprint.

> The `nx migrate` helps by automating the process of updating...

Reviewer: Why `...` at the end?
```suggestion
The `nx migrate` helps by automating the process of updating.
```

- package versions in your `package.json`
- configuration files (e.g. your Jest, ESLint or Nx config)

> - your source code (e.g. changes in imports)

Reviewer: ```suggestion
- your source code (e.g. fixing breaking changes or migrating to new best practices)
```

## How does it work?

Nx knows where its configuration files are and can therefore make sure they match the expected format or can alternatively adjust them. This automated update process, commonly referred to as "migration," becomes even more powerful when you leverage [Nx plugins](/packages). Nx plugins, which are NPM packages with a range of capabilities (code generation, task automation...), offer targeted updates based on their specific areas of responsibility.

For example, the [Nx ESLint plugin](/packages/linter) excels at configuring linting in your workspace. With its understanding of the configuration file locations, this plugin can provide precise migration scripts to update ESLint packages in your `package.json` and corresponding configuration files in your workspace when a new version is released.

Updating happens in three steps:

- The installed dependencies are updated including the `package.json` (and `node_modules`).

> - The source code in the repo is updated to match the new versions of packages in `package.json`.

Reviewer: ```suggestion
- The source code in the repo is updated to match the new versions of packages according to set of instructions specified in `migrations.json` file.
```

nx migrate latest # same as nx migrate nx@latest
```

> Note you can also specify an exact version by replacing `latest` with `nx@<version>`.

Reviewer: Do we want to mention that you can also update just a single third-party plugin with `nx migrate my-plugin@version`?

Author: yeah good idea, I'll add that

### Step 2: Running migrations

> You can now run the actual code migrations that were generated in the `migrations.json` in the previous step.

Reviewer: It's common to call code modification `codemods`, not migrations.

```suggestion
You can now run the actual codemods that were generated in the `migrations.json` in the previous step.
```

Author: yeah I know..but ppl might associate them to actual "codemods" (e.g. the Facebook lib). But I might add it in parentheses so ppl understand.

```shell
nx migrate --run-migrations
```

This will update your source code in your workspace in accordance with the implementation of the various migrations which ran and all the changes will be unstaged ready for you to review and commit yourself.

Note that each Nx plugin is able to provide a set of migrations which are relevant to particular versions of the package. Hence `migrations.json` will only contain migrations which are appropriate for the update you are currently applying.

### Step 3: Cleaning up

After you run all the migrations, you can remove `migrations.json` and commit any outstanding changes.

Note: You may want to keep the `migrations.json` until every branch that was created before the migration has been merged. Leaving the `migrations.json` in place allows devs to run `nx migrate --run-migrations` to apply the same migration process to their newly merged code as well.

## Need to opt-out of some migrations?

> Sometimes you need to temporarily opt-out from some migrations because your workspace is not ready yet. As mentioned, you can either manually adjust the `migrations.json` or run the update with the `--interactive` flag. More details in our [Advanced Update Process](/recipes/other/advanced-update) guide.

Reviewer: ```suggestion
Sometimes you need to temporarily opt-out from some migrations because your workspace is not ready yet. You can manually adjust the `migrations.json` or run the update with the `--interactive` flag to choose which migrations you accept. Find more details in our [Advanced Update Process](/recipes/other/advanced-update) guide.
```
